keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,947 @@
+
+### Goals
+
+- Kubelet should still work if images/containers are separated into different disks.
+  - Support writable layer being on same disk as Kubelet
+  - Images can be on the separate filesystem.
+
+### Possible Extensions in Post Alpha
+
+- Kubelet, Images and Containers on all separate disks.
+- Splitting images and containers across multiple disks.
+
+### Non-Goals
+
+- Multiple nodes can not share the same filesystem.
+- Separating kubelet data into different filesystems.
+
+## Proposal
+
+### User Deployment Options
+
+It is not a common pattern to separate the filesystems in most Kubernetes deployments.
+We will summarize the existing configurations that are possible today.
+
+#### Image File system and Node (Kubelet) FS same
+
+sda0: [writeable layer, emptyDir, logs, read-only layer]
+
+This is the default configuration for Kubernetes.  If container runtime is not configured in any special way, then NodeFS and ImageFS are assumed to be the same.
+
+If the node only has a nodefs filesystem that meets eviction thresholds, the kubelet frees up disk space in the following order:
+
+- Garbage collect dead pods and containers
+- Delete unused images
+
+#### Node FS and Image Filesystem separated
+
+sda0: [emptyDir, logs]
+sda1: [writeable layer, read-only layer]
+
+If the node has a dedicated imagefs filesystem for container runtimes to use, the kubelet does the following:
+
+- If the node filesystem meets the eviction thresholds, the kubelet garbage collects dead pods and logs
+- If the imagefs filesystem meets the eviction thresholds, the kubelet deletes all unused images and containers.
+- If ImageFS has disk pressure we will mark node as unhealthy and not allow new pods to be admitted until image disk pressure is gone.
+
+In case of disk pressure on each filesystem, what is garbage collected/stored on the disk?
+
+Node Filesystem:
+
+- Logs
+- Pods
+
+Image Filesystem:
+
+- Images
+- Containers
+
+CAdvisor detects the different disks based on mountpoints.  So if a user mounts a separate disk to /var/lib/containers or /var/lib/kubelet (which is default),
+Kubelet will think that the filesystem is split.
+
+#### Node And Writeable Layer on same disk while Images stored on separate disk
+
+sda0: [writable layer, emptyDir, logs]
+sda1: [read-only layer]
+
+A goal is to allow kubelet to have separate disks for read-only layer and everything else could be stored on the same disk as Kubelet.
+
+In case of disk pressure on each filesystem, what is garbage collected/stored on the disk?
+
+Node Fileystem:
+
+- Pods
+- Logs
+- Containers
+
+Image Filesystem:
+
+- Images
+
+Node Filesystem should monitor storage for containers in addition to ephemeral storage.
+
+### Container Runtime Configuration for CRIO
```

Comment:
I'm not sure this section is within scope of the enhancement. Seems more like the contents of an enablement blog or something

Reply:
Idk.  I wanted to at least talk a bit about this was enabled for CRIO.  Maybe it would be useful to have this in case of containerd support.  I can remove it but I found it useful to summarize this when I was writing.

Reply:
Idk.  I wanted to at least talk a bit about this was enabled for CRIO.  Maybe it would be useful to have this in case of containerd support.  I can remove it but I found it useful to summarize this when I was writing.

Reply:
I can take it or leave it honestly.  I can remove.

Reply:
removed.